### PR TITLE
Multi table insert

### PR DESCRIPTION
**Pull request by A-Baji**

### CODE REVIEW

**Review by jverswijver — APPROVED** [review]
Looks great! If you have extra time you should add a test for the form component in the testing suite.

File: pharus/dynamic_api_gen.py [hunk 1]
```diff
@@ -24,25 +24,26 @@ def populate_api():
 """
     route_template = """
 
-@app.route('{route}', methods=['GET'])
+@app.route('{route}', methods=['GET', 'POST'])
 @protected_route
 def {method_name}(jwt_payload: dict) -> dict:
 
-    if request.method in {{'GET'}}:
+    if request.method in {{'GET', 'POST'}}:
```
Comment (jverswijver) [inline pharus/dynamic_api_gen.py:31]: To fix the tests split the If into an if-elseif for `post` and `get`, in the get route do not do `payload=request.get_json()` just do not set it and set the default to `none` in the `component_interface`.

Looking at the failed tests it seems that `request.get_json()` produces an error when being used in a get route.

**Review by jverswijver — CHANGES_REQUESTED** [review]
Sorry did not see that the tests did not pass. I spent some time looking at it and I think I know what the problem is, I will highlight it in this review.

File: pharus/component_interface.py [hunk 2]
```diff
@@ -74,19 +73,39 @@ def __init__(self, name, component_config, static_config, jwt_payload: dict):
...
-                connection=dj.conn(
-                    host=jwt_payload["databaseAddress"],
-                    user=jwt_payload["username"],
-                    password=jwt_payload["password"],
-                    reset=True,
-                ),
-            )
-            for s in inspect.getfullargspec(self.dj_query).args
-        ]
+        if "dj_query" in component_config:
```
Comment (guzman-raphael) [inline pharus/component_interface.py:76]: Think it would be best to split into a separate fundamental component. We can perhaps rename this one to be `FetchComponent` and create a new base `InsertComponent`.

File: pharus/component_interface.py [hunk 3]
```diff
@@ -74,19 +73,39 @@ def __init__(self, name, component_config, static_config, jwt_payload: dict):
...
+            exec(component_config["dj_query"], globals(), lcls)
+            self.dj_query = lcls["dj_query"]
+            self.vm_list = [
```
Comment (guzman-raphael) [inline pharus/component_interface.py:79]: This should be named `self.tables` to be more clear on what it represents. Also from discussion, we'll need to add a `connection` attribute to be accessible for transactions.

File: pharus/component_interface.py [hunk 4]
```diff
@@ -74,19 +73,39 @@ def __init__(self, name, component_config, static_config, jwt_payload: dict):
...
+                        host=jwt_payload["databaseAddress"],
+                        user=jwt_payload["username"],
+                        password=jwt_payload["password"],
+                        reset=True,
+                    ),
+                )
+                for s in inspect.getfullargspec(self.dj_query).args
+            ]
+        if "tables" in component_config:
+            self.tables = component_config["tables"]
```
Comment (guzman-raphael) [inline pharus/component_interface.py:93]: No need to store the config after we've loaded it.

File: pharus/component_interface.py [hunk 5]
```diff
@@ -114,6 +133,15 @@ def restriction(self):
             ]
         )
 
+class InsertComponent(QueryComponent):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def dj_query_route(self):
+        [t.insert1(
```
Comment (guzman-raphael) [inline pharus/component_interface.py:141]: We'll need to put this into a [transaction](https://docs.datajoint.org/python/manipulation/4-Transactions.html) context. Also, for functions that do not return data, it is atypical to include into a list. It is tempting to put them into a list comprehension and I used to do this too. :sweat_smile:. It started becoming difficult to review after some time...

File: pharus/component_interface.py [hunk 6]
```diff
@@ -326,4 +354,5 @@ def dj_query_route(self):
     "file:image:attach": FileImageAttachComponent,
     "slider": BasicQuery,
     "dropdown-query": BasicQuery,
+    "form": InsertComponent,
```
Comment (guzman-raphael) [inline pharus/component_interface.py:357]: This comment is for the `basicquery` line above. (GH wouldn't let me comment directly on it)

We can simplify the implementation of it to just point to `FetchComponent` and move the `dj_query_route` into `FetchComponent` since it is overloaded by each child class.

File: pharus/dynamic_api_gen.py [hunk 7]
```diff
@@ -24,25 +24,26 @@ def populate_api():
 """
     route_template = """
 
-@app.route('{route}', methods=['GET'])
+@app.route('{route}', methods=['GET', 'POST'])
```
Comment (guzman-raphael) [inline pharus/dynamic_api_gen.py:27]: I'd suggest we only expose the REST verb if there is a valid path there with the changes proposed. Currently, this would expose `POST` endpoints for `FetchComponent` routes and `GET` for `InsertComponent` routes.

We can either make this a new variable or make a new template.

File: tests/init/test_dynamic_api_spec.yaml [hunk 8]
```diff
@@ -15,6 +15,16 @@ SciViz: # top level tab
         grid1:
           type: fixed
           components:
+            insert:
```
Comment (guzman-raphael) [inline tests/init/test_dynamic_api_spec.yaml:18]: Would you add test(s) to verify the behavior?

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@A-Baji Nice job in figuring a lot of this out. :muscle: 

Based on this, we can tweak our structure a bit to properly accommodate.

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@A-Baji Looking good. :sunglasses:

We should add a `GET` route for the fields info. This is necessary so we can map between the display and what the database needs. I've updated the issue with [specifications](https://github.com/datajoint/pharus/issues/124#issuecomment-1217282872). Could you look into that next?

File: pharus/component_interface.py [hunk 9]
```diff
@@ -114,8 +109,167 @@ def restriction(self):
...
+                    field_datatype = datatype_map[field["destination"]]
+                    fields.append(
+                        dict(type=field_type, datatype=field_datatype, name=field_name)
+                    )
+                elif field_type == "table":
+                    s, t = field["destination"].split(".")
+                    if "map" in field:
+                        proj_map = {
+                            attr["input"]
+                            if "input" in attr
```
Comment (guzman-raphael) [inline pharus/component_interface.py:211]: This deep nested if's are a bit hard to follow the logic. A minor point but still took me a bit to understand it on first glance.

File: pharus/component_interface.py [hunk 10]
```diff
@@ -114,8 +109,167 @@ def restriction(self):
...
+                    else:
+                        records = getattr(
+                            dj.VirtualModule(
+                                s,
+                                s,
+                                connection=self.connection,
+                            ),
+                            t,
+                        ).fetch("KEY")
+                    fields.append(
```
Comment (guzman-raphael) [inline pharus/component_interface.py:236]: `append`'s are slow and a memory expensive operation.

File: pharus/component_interface.py [hunk 11]
```diff
@@ -114,8 +109,167 @@ def restriction(self):
...
+                        dict(type=field_type, values=records, name=field_name)
+                    )
+        else:
+            for t in self.tables:
+                for v in t.heading.attributes.values():
+                    is_fk = False
```
Comment (guzman-raphael) [inline pharus/component_interface.py:242]: Flags like this are unnecessary if you use a variable's *truthy* characteristic.

File: pharus/component_interface.py [hunk 12]
```diff
@@ -114,8 +109,167 @@ def restriction(self):
...
+                    for p in t.parents(as_objects=True):
```
Comment (guzman-raphael) [inline pharus/component_interface.py:243]: You can identify the unique parents for the entire group of insert tables before iterating over each so you only do this on unique ones. Additionally, it is not necessary to iterate over every parent **and** over every attribute.

File: pharus/component_interface.py [hunk 13]
```diff
@@ -114,8 +109,167 @@ def restriction(self):
...
+                self.payload = {field_name_map[k]: v for k, v in self.payload.items()}
+            for t in self.tables:
+                t.insert1(
+                    {k: v for k, v in self.payload.items() if k in t.heading.attributes}
+                )
+        return "Insert successful"
+
+    def fields_route(self):
+        fields = []
+        if self.fields_map:
```
Comment (guzman-raphael) [inline pharus/component_interface.py:194]: Large `if` blocks like this can be a bit hard to follow the logic.

File: pharus/component_interface.py [hunk 14]
```diff
@@ -114,8 +109,167 @@ def restriction(self):
...
+                for v in t.heading.attributes.values():
+                    is_fk = False
+                    for p in t.parents(as_objects=True):
+                        if v.name in p.primary_key:
+                            is_fk = True
+                            field_name = (
+                                f"{p.database}.{dj.utils.to_camel_case(p.table_name)}"
+                            )
+                            if field_name not in [
+                                field["name"] for field in fields
```
Comment (guzman-raphael) [inline pharus/component_interface.py:250]: `field_name` seems incorrectly named but more importantly `fields` seems a bit ambiguous to me. I'm having a hard time seeing where it get its value from. :thinking:

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@A-Baji Great job! :+1: Can see a well thought-out approach here.

Have some additional points for consideration in your algorithm approach. I actually have an example to share that can help you improve on all of these. :nerd_face: 

Oh, and what if I could tell you we could get the following as a bonus too?
- Allows multiple inserts to be submitted
- Configuration can omit individual maps if no change
- Configuration can omit `map` from `table` maps if no change
- slightly less code

Find me and we can knock this out in a tagup together.

File: pharus/component_interface.py (hunk 14 above)
Comment (guzman-raphael) [reply pharus/component_interface.py]: Ah, the `append` use explains how this is being set.

**Review by guzman-raphael — APPROVED** [review]

**Review by jverswijver — APPROVED** [review]
